### DISCUSSION TITLE
Invalid diff shown in github UI

### DISCUSSION THREAD

**Discussion by markdroth** [community/community#20178] [Pull Requests]
I am trying to look at the changes in the latest commit on a PR, but the diff that github is showing me is just plain invalid.  Here's the page I'm looking at:

https://github.com/grpc/grpc/pull/30008/files/2c3b9a2cc210ab5b93fd7b600a1edad92d6523f6..b59518500f78aae8bcc9e9bbe9f811a7feacae2a

Here's what I see for the first file in the diff:

![image](https://user-images.githubusercontent.com/18664614/177863633-c36940c1-1865-4b52-8173-56b074989815.png)

Here's the actual diff that I see when I use "git diff" locally:

```
@@ -712,7 +713,7 @@ class CdsLbFactory : public LoadBalancingPolicyFactory {
  public:
   OrphanablePtr<LoadBalancingPolicy> CreateLoadBalancingPolicy(
       LoadBalancingPolicy::Args args) const override {
-    auto xds_client = args.args.GetObjectRef<XdsClient>();
+    auto xds_client = args.args.GetObjectRef<GrpcXdsClient>();
     if (xds_client == nullptr) {
       gpr_log(GPR_ERROR,
               "XdsClient not present in channel args -- cannot instantiate "
```

Note that the line that actually changed here is line 715.  However, the github UI has highlighted line 716 as red/green instead of line 715, even though there is no change on that line.  And even worse, it is showing the wrong text for line 715 on the right hand side of the diff: that line actually shows duplicate text from line 714 rather than showing the actual new text on line 715.

That's just one example; this particular github page has a number of files whose diffs are displayed in this invalid way.

I'm guessing that there's some sort off-by-one bug in the UI rendering somewhere, but this problem is making it very hard to rely on the github UI.  Can someone please investigate this bug?

Thanks!

**Comment by pavadeli** [comment]
I get the impression this is introduced with the recent "deferred syntax highlighting" feature: https://github.blog/changelog/2022-06-24-deferred-syntax-highlighting/

When throttling the network, I get a correct diff at first (after the initial loading of the page):
<img width="815" alt="image" src="https://user-images.githubusercontent.com/5168053/178237226-d0c761b2-e4e2-4674-bf7b-1b5db8bc5f29.png">

It then shifts/gets corrupted afterwards while applying the syntax highlighting lazily:
<img width="728" alt="image" src="https://user-images.githubusercontent.com/5168053/178237348-4f10cc71-72ae-44f7-83fe-007094845e40.png">

**Reply by pavadeli** [reply]
~If you require a workaround and if you can live without syntax highlighting, you can use request-blocking by browser dev-tools to block all requests containing `deferred-diff-lines-element`. This effectively breaks the deferred syntax highlighting and seems to prevent the issue for now.~

Scratch that, that only works for files that loaded on initial page-load. These "initial files" are loaded without syntax highlighting and then lazily upgraded to have syntax-highlighting afterwards. On larger PRs most files are omitted from inital page-load. These seem to load (deferred) with syntax-highlighting included and are corrupt on arival.

**Reply by galpHub** [reply]
Is the droid you are looking for @pavadeli ?
https://stackoverflow.com/questions/55479729/github-pull-request-shows-wrong-diff

**Reply by pavadeli** [reply]
> Is the droid you are looking for @pavadeli ? https://stackoverflow.com/questions/55479729/github-pull-request-shows-wrong-diff

These are not the droids I am looking for. The stackoverflow issue is describing unexpected files in a diff. The bug here (which seems to be fixed now?) was an off-by-one bug at GitHub which corrupted a single file in the GitHub UI while looking at it (literally, you could see it get corrupted when syntax highlighting kicked in).

**Comment by pavadeli** [comment]
It seems to be fixed now, at least for the link @markdroth posted.
